src/rust/bitbox02-rust/src/backup.rs

```diff
@@ -0,0 +1,357 @@
+
+            let checksum = compute_checksum(
+                content.metadata.as_ref().unwrap(),
+                &backup_data.0,
+                content.length,
+            )?;
+            if checksum != content.checksum {
+                Err(())
+            } else {
+                Ok((backup_data, content.metadata.unwrap()))
+            }
+        }
+        _ => Err(()),
+    }
+}
+
+/// Does a bitwise majority vote to recover the contents of potentially corrupted data. All three
+/// buffers be of the same length.
+fn bitwise_recovery(buf1: &[u8], buf2: &[u8], buf3: &[u8]) -> Result<Zeroizing<Vec<u8>>, ()> {
+    let len = buf1.len();
+    if len != buf2.len() || len != buf3.len() {
+        return Err(());
+    }
+    let mut recovered_contents = Zeroizing::new(vec![0u8; len]);
+    for i in 0..len {
+        for bit in 0..8 {
+            let bit1 = buf1[i] & (1 << bit);
+            let bit2 = buf2[i] & (1 << bit);
+            let bit3 = buf3[i] & (1 << bit);
+            let winner = if bit1 == bit2 || bit1 == bit3 {
+                bit1
+            } else {
+                bit2
+            };
+            recovered_contents[i] |= winner;
+        }
+    }
+    Ok(recovered_contents)
+}
+
+pub fn load(dir: &str) -> Result<(Zeroizing<BackupData>, pb_backup::BackupMetaData), ()> {
+    let files = bitbox02::sd::list_subdir(Some(dir))?;
+    if files.len() != 3 {
+        return Err(());
+    }
+    let file_contents: [Zeroizing<Vec<u8>>; 3] = [
+        bitbox02::sd::load_bin(&files[0], dir)?,
+        bitbox02::sd::load_bin(&files[1], dir)?,
+        bitbox02::sd::load_bin(&files[2], dir)?,
+    ];
+    for contents in file_contents.iter() {
+        if let o @ Ok(_) = load_from_buffer(contents) {
+            return o;
+        }
+    }
+    // If we arrived here, it means all three copies of the backup are corrupted (couldn't decode or
+    // failed the checksum verification). We try to recover with a bit-wise majority vote using the
+    // three copies of the backup. This only works if all three files have the same size.
+    load_from_buffer(&bitwise_recovery(
+        &file_contents[0],
+        &file_contents[1],
+        &file_contents[2],
+    )?)
+}
+
+static GENERATOR: &str = {
+    let generator = env!("FIRMWARE_VERSION_SHORT");
+    // Need explicit check as the env var could be set to an empty string accidentally (env!() only
+    // panics if it is not set at all).
+    if generator.is_empty() {
+        panic!("FIRMWARE_VERSION_SHORT is not set");
+    }
+    generator
+};
+
+pub fn create(
+    seed: &[u8],
+    name: &str,
+    backup_create_timestamp: u32,
+    seed_birthdate_timestamp: u32,
+) -> Result<(), Error> {
+    let backup_data = zeroize::Zeroizing::new(BackupData(Box::new(pb_backup::BackupData {
+        seed_length: seed.len() as _,
+        seed: padded_seed(seed).to_vec(),
+        birthdate: seed_birthdate_timestamp,
+        generator: GENERATOR.into(),
```

Comment:
need to check if the C version serialized this as a zero terminated string. if so, we must do the same for backwards compatibility.

Reply:
Luckily it seems that the strings were encoded as normal protobuf strings and not some weird zero terminated or padded string:

https://github.com/digitalbitbox/bitbox02-firmware/pull/956

Also the fixture (written by nanopb) decoded:

```
In [1]: bytes.fromhex('0a6c0a6a0a2017834e53e17370800c0bc49b49ef3f1309df104d7239d
   ...: b5bbd093c90eefc995112110891bec6fb0512094d7920426974426f782233081012208af
   ...: 64d31126a39b98f59708a3a463e5b000000000000000000000000000000001891bec6fb0
   ...: 5220776392e31332e30')
Out[1]: b'\nl\nj\n \x17\x83NS\xe1sp\x80\x0c\x0b\xc4\x9bI\xef?\x13\t\xdf\x10Mr9\xdb[\xbd\t<\x90\xee\xfc\x99Q\x12\x11\x08\x91\xbe\xc6\xfb\x05\x12\tMy BitBox"3\x08\x10\x12 \x8a\xf6M1\x12j9\xb9\x8fYp\x8a:F>[\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x18\x91\xbe\xc6\xfb\x05"\x07v9.13.0'
```

```
$ echo 0a6c0a6a0a2017834e53e17370800c0bc49b49ef3f1309df104d7239db5bbd093c90eefc995112110891bec6fb0512094d7920426974426f782233081012208af64d31126a39b98f59708a3a463e5b000000000000000000000000000000001891bec6fb05220776392e31332e30 | xxd -r -p | protoc --decode_raw
1 {
  1 {
    1: "\027\203NS\341sp\200\014\013\304\233I\357?\023\t\337\020Mr9\333[\275\t<\220\356\374\231Q"
    2 {
      1: 1601281809
      2: "My BitBox"
    }
    4 {
      1: 16
      2: "\212\366M1\022j9\271\217Yp\212:F>[\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000"
      3: 1601281809
      4: "v9.13.0"
    }
  }
}
```

one can clearly see that the name and generator are not null terminated.